### PR TITLE
AP_BattMonitor: added SUI SMBUS battery backend

### PR DESCRIPTION
**Pull request by tridge**
originally by Kelly Schrock

This also adds a BUS param to allow for SMBUS interfaces on any i2c bus

### CODE REVIEW

File: libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp [hunk 1]
```diff
@@ -162,6 +162,13 @@ const AP_Param::GroupInfo AP_BattMonitor_Params::var_info[] = {
     // @User: Advanced
     AP_GROUPINFO("ARM_MAH", 19, AP_BattMonitor_Params, _arming_minimum_capacity, 0),
 
+    // @Param: BUS
+    // @DisplayName: I2C bus number
+    // @Description: This specifies the I2C bus number for SMBus based backends
+    // @Values: 0:Bus0(internal),1:Bus1(external),2:Bus2(auxillary),3:Bus3(auxillary2)
+    // @User: Advanced
+    AP_GROUPINFO("BUS", 20, AP_BattMonitor_Params, _bus, 0),
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_Params.cpp]: Not a deal breaker, but ideally this would start going into the subgroup of parameters for a specific backend type so that we could start hiding these from the generic battery backends.

Side note: Do we want to reserve a value that simply tries all the buses till it finds a matching instance to ease the general setup? Past that I'm quite happy for this to be added as it fixes a couple of long outstanding pain points with SMBus batteries.

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp [hunk 2]
```diff
@@ -113,13 +114,27 @@ AP_BattMonitor::init()
...
+                                                                                          100000, true, 20), 3);
+                break;
+            case AP_BattMonitor_Params::BattMonitor_TYPE_Endurance:
+                _params[instance]._bus.set_default(AP_BATTMONITOR_SMBUS_BUS_INTERNAL),
+                drivers[instance] = new AP_BattMonitor_WithCellCount(*this, state[instance], _params[instance],
+                                                                     hal.i2c_mgr->get_device(_params[instance]._bus, AP_BATTMONITOR_SMBUS_I2C_ADDR,
+                                                                                             100000, true, 20), 6);
+                break;
             case AP_BattMonitor_Params::BattMonitor_TYPE_MAXELL:
+                _params[instance]._bus.set_default(AP_BATTMONITOR_SMBUS_BUS_EXTERNAL),
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor.cpp]: This is really messy, I get that the maxell driver wanted the external by default, but this sorta magic twiddling is a bit confusing, and can lead to people accidentally. I think a default param to scan all buses would probably be less painful at that point.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    // read bytes
+    if (!_dev->read_registers(reg, data, max_len)) {
+        return 0;
+    }
+
+    // return success
+    return max_len;
+}
+
+AP_BattMonitor_WithCellCount::AP_BattMonitor_WithCellCount(AP_BattMonitor &mon,
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: What is this wrapper buying you? Just seems to make it more confusing, why not just instantiate the SUI one directly?

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+                _state.cell_voltages.cells[i] = cell;
+                pack_voltage_mv += (float)cell;
+            }
+
+            if(_cell_count >= 4) {
+                float avg = (pack_voltage_mv / 4);
+
+                for(uint8_t i = 4; i < 6; ++i) {
+                    _state.cell_voltages.cells[i] = avg;
+                    pack_voltage_mv += _state.cell_voltages.cells[i];
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: We shouldn't report a fake value here, we should be forcing a read of the actual pack voltage.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+            float pack_voltage_mv = 0.0f;
+
+            for (uint8_t i = 0; i < MIN(4, _cell_count); i++) {
+                uint16_t cell = voltbuff[(i * 2) + 1] << 8 | voltbuff[i * 2];
+                _state.cell_voltages.cells[i] = cell;
+                pack_voltage_mv += (float)cell;
+            }
+
+            if(_cell_count >= 4) {
+                float avg = (pack_voltage_mv / 4);
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: Since we have to force a pack read anyways, just take the (actual voltage - accumulated cell voltage) / 2 to get a better and more accurate value.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+                                                   uint8_t num_cells)
+    : AP_BattMonitor_SMBus_SUI(mon, mon_state, params, std::move(dev), num_cells)
+{}
+
+void AP_BattMonitor_WithCellCount::timer() {
+    AP_BattMonitor_SMBus_SUI::timer();
+    read_remaining_capacity();
+}
+
+void AP_BattMonitor_WithCellCount::read_cell_voltages() {
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: What actually calls this? The SUI code will call the SUI specific one not this...

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                    AP_BattMonitor_Params &params,
+                    AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev,
+                    uint8_t cell_count
+                    );
+
+    void init(void) override;
+
+protected:
+
+    std::vector<float> currents;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h]: Are we actually okay with letting `std::vector` in here?

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+        return;
+    }
+
+    // read current
+    if (read_block_bare(REG_CURRENT, buff, 4, false)) {
+        float current_amps = -(float)((int32_t)((uint32_t)buff[3]<<24 | (uint32_t)buff[2]<<16 | (uint32_t)buff[1]<<8 | (uint32_t)buff[0])) / 1000.0f;
+
+        currents.push_back(current_amps);
+        if(currents.size() > 3) {
+            currents.erase(currents.begin());
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: Doesn't this make the average divide by 0 right after this?

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+
+bool AP_BattMonitor_SMBus_SUI::read_remaining_capacity() {
+    int32_t capacity = _params._pack_capacity;
+
+    if (capacity > 0) {
+        uint16_t data;
+        if (read_word(BATTMONITOR_SMBUS_REMAINING_CAPACITY, data)) {
+            float consumed = (capacity - data);
+            if((consumed > 0) && (consumed < capacity)) {
+                consumed_mahs.push_back(consumed);
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: Why? The battery monitor's job in life is to do this for us.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+    }
+}
+
+bool AP_BattMonitor_SMBus_SUI::read_temp(void)
+{
+    uint16_t data;
+    if (read_word(BATTMONITOR_SMBUS_TEMP, data)) {
+        float temp = ((float)(data - 2731)) * 0.1f;
+
+        if((fabs(temp - avg(temperatures)) < 100) || (temperatures.size() == 0 && temp < 1000)) {
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: How is this noisy enough to be worth filtering?

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp (hunk 2 above)
Comment (JoshWelsh) [reply libraries/AP_BattMonitor/AP_BattMonitor.cpp]: Can lead to people accidentally...what?

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: removed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: done

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 6 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: fixed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 8 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: fixed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 10 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: fixed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 9 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: fixed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h (hunk 7 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h]: removed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 5 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: yep, done

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+    if (read_word(BATTMONITOR_SMBUS_TEMP, data)) {
+        float temp = ((float)(data - 2731)) * 0.1f;
+        _state.temperature = temp;
+        _state.temperature_time = AP_HAL::millis();
+    }
+
+    return true;
+}
+
+bool AP_BattMonitor_SMBus_SUI::read_remaining_capacity()
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: Again, looks be the mostly same as the one from `AP_BattMonitor_SMBus` ?

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+
+    // copy data (excluding PEC)
+    memcpy(data, &buff[1], bufflen);
+
+    // return success
+    return true;
+}
+
+// read_bare_block - returns true if successful
+bool AP_BattMonitor_SMBus_SUI::read_block_bare(uint8_t reg, uint8_t* data, uint8_t len) const
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: Whoa, why is this needed? This looks to give up on all the PEC protection, which doesn't make sense...

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+    if (phase_voltages) {
+        read_cell_voltages();
+        read_remaining_capacity();
+        return;
+    }
+
+    // read current
+    int32_t current_ma;
+    if (read_block_bare(REG_CURRENT, (uint8_t *)&current_ma, sizeof(current_ma))) {
+        _state.current_amps = -current_ma * 0.001;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: Micro optimization, but the compiler does better with `current_ma * -0.001;` saves 4 bytes of flash, and one less floating point operation.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+    }
+
+    // read the manufacturers data for button press
+    uint8_t mf_data[6];
+    if (read_block(BATTMONITOR_SMBUS_MANUFACTURE_DATA, mf_data, sizeof(mf_data))) {
+        bool pressed = (mf_data[1] >> 3) & 0x01;
+
+        if (button_press_count >= BATTMONITOR_SMBUS_SUI_BUTTON_DEBOUNCE) {
+            // battery will power off
+            AP_Notify::flags.powering_off = true;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: This doesn't set the notify object directly anymore, it should be setting `_state.is_powering_off` instead. You should also test that this actually detects the notification correctly from the button on the battery, this is a copy of the Solo implementation so I'm not sure it's actually been verified as being a reachable/correct bit of code here.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+        } else if (pressed) {
+            // battery will power off if the button is held
+            button_press_count++;
+
+        } else {
+            // button released, reset counters
+            button_press_count = 0;
+            AP_Notify::flags.powering_off = false;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: Again this should manipulate the state object not the notify one. In practice the notify one is considered a one way gate (primarily for the multiple battery case).

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+        _state.current_amps = -current_ma * 0.001;
+        _state.last_time_micros = tnow;
+    }
+
+    read_full_charge_capacity();
+
+    read_remaining_capacity();
+
+    // timeout after 5 seconds
+    if ((tnow - _state.last_time_micros) > AP_BATTMONITOR_SMBUS_TIMEOUT_MICROS) {
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: This is pointless here, you never return before you get here, you need to have a check that you actually managed to do . the capacity calculations before this, otherwise the driver might as well always be healthy and just not manipulate this.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+        _state.cell_voltages.cells[i] = cell;
+        pack_voltage_mv += (float)cell;
+    }
+
+    if (cell_count >= SUI_MAX_CELL_READ) {
+        // we can't read voltage of all cells. get overall pack voltage to work out
+        // an average for remaining cells
+        uint16_t total_mv;
+        if (read_block(BATTMONITOR_SMBUS_VOLTAGE, (uint8_t *)&total_mv, sizeof(total_mv)) &&
+            total_mv >= pack_voltage_mv) {
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: If `total_mv` isn't greater we have a lot of problems, maybe worth removing the check and allowing the voltages to go crazy as at least it flags a problem?

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+            for (uint8_t i = SUI_MAX_CELL_READ; i < cell_count; i++) {
+                _state.cell_voltages.cells[i] = (total_mv - pack_voltage_mv) / (cell_count - SUI_MAX_CELL_READ);
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: Lifting this out to a const `uint16_t` before the for loop saves 12 bytes of flash, and should be faster.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+            }
+            pack_voltage_mv = total_mv;
+        } else {
+            // we can't get total pack voltage. Use average of cells we have so far
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: I'm not really a fan of this code path if the read_register worked, and even in the case where the read failed it can result in more battery fluctuation from one read to the other based on if the read worked, but I guess it's better then giving up forever?

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 20]
```diff
@@ -0,0 +1,228 @@
...
+                pack_voltage_mv += _state.cell_voltages.cells[i];
+            }
+        }
+    }
+
+    _has_cell_voltages = true;
+
+    // accumulate the pack voltage out of the total of the cells
+    _state.voltage = pack_voltage_mv * 0.001;
+    last_volt_read_us = AP_HAL::micros();
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: This is a write only member, what's it trying to do for you?

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+                             AP_BattMonitor::BattMonitor_State &mon_state,
+                             AP_BattMonitor_Params &params,
+                             AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev,
+                             uint8_t cell_count
+                            );
+
+    void init(void) override;
+
+private:
+    float currents[SUI_MAXCELLS];
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h]: These arrays were for the old averaging code, and should be removed.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+    bool read_block(uint8_t reg, uint8_t* data, uint8_t len) const;
+    bool read_block_bare(uint8_t reg, uint8_t* data, uint8_t len) const;
+
+    // average over one of currents, consumed_mahs, temperatures
+    float avg(const float *p) const;
+
+    uint8_t button_press_count;
+
+    const uint8_t cell_count;
+    int32_t capacity;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h]: No one seems to use this, what was it's purpose?

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 23]
```diff
@@ -0,0 +1,228 @@
...
+    _state.healthy = true;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: This is inconsistent with the use of the timer/setting this to true/false in the outside calling code, this needs to be consistent with that for update time out periods.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h [hunk 24]
```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <AP_Common/AP_Common.h>
+#include <AP_Param/AP_Param.h>
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h]: No reason to include AP_Param.h here that I can see? I know it's coming in transitively via other headers, but we don't actually need it, no reason to make it harder to fix the other headers then it already is...

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 25]
```diff
@@ -0,0 +1,228 @@
...
+    // read current
+    int32_t current_ma;
+    if (read_block_bare(REG_CURRENT, (uint8_t *)&current_ma, sizeof(current_ma))) {
+        _state.current_amps = -current_ma * 0.001;
+        _state.last_time_micros = tnow;
+    }
+
+    read_full_charge_capacity();
+
+    read_remaining_capacity();
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: You do this in both phases, why?

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 26]
```diff
@@ -0,0 +1,228 @@
...
+    if (_dev && timer_handle) {
+        // run twice as fast for two phases
+        _dev->adjust_periodic_callback(timer_handle, 50000);
+    }
+}
+
+/*
+  average over an array of the number of readable cells
+ */
+float AP_BattMonitor_SMBus_SUI::avg(const float *p) const
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: done

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 27]
```diff
@@ -0,0 +1,228 @@
...
+        }
+    }
+
+    read_temp();
+
+    read_serial_number();
+    read_remaining_capacity();
+}
+
+bool AP_BattMonitor_SMBus_SUI::read_temp(void)
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: removed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 13 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: changed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 20 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: fixed with update_health() method

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h (hunk 22 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h]: removed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 23 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: fixed with update_health()

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h (hunk 24 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h]: fixed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 25 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: fixed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h (hunk 21 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.h]: removed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 11 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: removed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 16 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: fixed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 17 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: I don't like the idea of making it go crazy. I've changed it so it reads zero volts for the extra cells. That should be distinctive enough in the logs

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 19 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: yes, I think we need to make best effort

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp [hunk 28]
```diff
@@ -0,0 +1,164 @@
...
+#include <AP_Common/AP_Common.h>
+#include <AP_Math/AP_Math.h>
+#include <AP_Notify/AP_Notify.h>
+#include "AP_BattMonitor.h"
+#include "AP_BattMonitor_SMBus_SUI.h"
+
+extern const AP_HAL::HAL& hal;
+
+// button held down for 3 intervals will cause a power off event
+#define BATTMONITOR_SMBUS_SUI_BUTTON_DEBOUNCE 3
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: removed

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp (hunk 18 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp]: done

**Comment by rmackay9** [comment]
By the way, Maxell batteries also suffer from an issue in which the reads of individual cell voltages fails a lot so I'd be happy with the Maxell battery driver to be changed to a similar structure... or perhaps we could use a state variable instead of the slightly more complex method of changing the timer frequency.

**Comment by tridge** [comment]
@WickedShell ping?

**Comment by rmackay9** [comment]
I'm tentatively planning on including this in Copter-4.0.4-rc1 as a prerequisite to including the NeoDesign battery drivery.
